README.md
A library for the Raspberry Pi Sense HAT LED Screen
====================================================

[![crates.io](https://img.shields.io/crates/v/sensehat-screen.svg)](https://crates.io/crates/sensehat-screen)
[![docs](https://docs.rs/sensehat-screen/badge.svg)](https://docs.rs/sensehat-screen)
[![Build Status](https://travis-ci.org/saibatizoku/sensehat-screen-rs.svg?branch=master)](https://travis-ci.org/saibatizoku/sensehat-screen-rs)


The [Raspberry Pi Sense HAT](https://www.raspberrypi.org/products/sense-hat/) has an 8×8 RGB LED matrix that provides its own driver for the Linux framebuffer.

This library provides a thread-safe, strong-typed, high-level API for the LED matrix, treating it as you would any other screen on a Linux box.

# Requirements

This crate supports Rust stable version v1.32.0 and higher, and is tested on nightly continuously.

# Changelog

For a detailed summary of how changes are incorporated into the code, we keep a [CHANGELOG](./CHANGELOG.md).

# Usage

To use this crate with the default features, add this to your `Cargo.toml`:
```cargo
[dependencies]
sensehat-screen = "0.2"
```

or, to manually specify the features::

```cargo
[dependencies]
sensehat-screen = { version = "0.2", default-features = false, features = ["fonts"] }
```

Then you can use it with your crate:

```rust
extern crate sensehat_screen

use sensehat_screen::{FrameLine, PixelColor, Screen, FONT_COLLECTION};
```

# Documentation

You can find the documentation for the latest release at: [https://docs.rs/sensehat-screen](https://docs.rs/sensehat-screen).

It contains a general description of the types contained in the library, it is a good to place to get familiar with the methods available.


# Examples

## Source Code Examples

You can find working examples in the source code.

* [Blinking RGB shapes](./examples/blink.rs)

  This example builds low-level `FrameLine`s by hand, writes them
  to the `Screen` at set intervals.

* [A letter from ゆにち (Yunichi) to Toño](./examples/letter.rs)

  This example makes use of the built-in 8x8 `FontCollection`. The collection is used to sanitize a user-provided `&str`, returning a `FontString` that includes only those characters provided by the FontCollection (basically, the full `latin` set, some `greek`, some `hiragana`, and legacy `ascii box` and `ascii block` sets).

  Each font is given stroke color of 50% white, `PixelColor::WHITE.dim(0.5)`, and then it is converted into a `FrameLine`, which is finally written to the `Screen`.

  See the [font8x8](https://github.com/saibatizoku/font8x8-rs) crate for more details.

* [Rotate demo](./examples/rotate.rs)

  This example makes use of `PixelFrame`, an ergonomic wrapper for representing each pixel in the 8x8 LED Matrix: `[PixelColor; 64]`.

  Like the `letter` example, it makes use of the built-in `FontCollection`, to create two hand-made `PixelFrame`s, made from the a yellow-colored `Ñ`, and a magenta-colored `ó`.

  Each frame is then subject to counter-clockwise rotation in steps of 90°.

* [Offset demo](./examples/offset.rs)

  Starts with an empty screen, and two font-symbols. Each symbol is then made to slide-in from the left by setting an `Offset` progressively.

  Once the symbol is fully displayed, it slides-out to the right, again by incrementing the offset, until the symbol disappears.

  The same process is done with the top-bottom offset directions.

* [Clip demo](./examples/clip.rs)

  A `Clip` is a `PixelFrame` that is the result of merging two `PixelFrame`. They are useful for manually constructing sequences of frames. Generally, you will prefer to work with `Scroll`, and `FrameSequence` iterators.

* [Scroll - Top To Bottom](./examples/scroll-top-bottom.rs)

  The name says it all: it builds a scroll of frames, and displays it from top to bottom, one clipped-frame at a time.

  Makes use of `FontString`, `Scroll`, and `FrameSequence` to work.

* [Scroll - Bottom To Top](./examples/scroll-bottom-top.rs)

  The name says it all: it builds a scroll of frames, and displays it from bottom to top, one clipped-frame at a time.

  Makes use of `FontString`, `Scroll`, and `FrameSequence` to work.

* [Scroll - Left To Right](./examples/scroll-left-right.rs)

  The name says it all: it builds a scroll of frames, and displays it from left to right, one clipped-frame at a time.

  Makes use of `FontString`, `Scroll`, and `FrameSequence` to work.

* [Scroll - Right To Left](./examples/scroll-right-left.rs)

  The name says it all: it builds a scroll of frames, and displays it from right to left, one clipped-frame at a time.

  Makes use of `FontString`, `Scroll`, and `FrameSequence` to work.

## A Simple, Low-Level Example

The following program shows how to:

* Open the framebuffer file-descriptor for the LED matrix screen (`screen`)
* Define a pixel color (`red_pixel`)
* Define a slice of pixel colors that represents the screen (`all_64_pixels`)
* Turn that slice into a valid pixel frame
* Show the frame on the screen

```rust
extern crate sensehat_screen;

use sensehat_screen::{PixelFrame, PixelColor, Screen};

fn main() {
    let mut screen = Screen::new("/dev/fb1")
        .expect("Could not open the framebuffer for the screen");

    let red_pixel = PixelColor::new(255, 0, 0); // The pixel color's RGB components are each in the range of 0 <= c < 256.

    let all_64_pixels = &[red_pixel; 64];   // A single vector of 8 x 8 = 64 pixel colors (rows are grouped by chunks of 8)

    let all_red_screen = PixelFrame::from_pixels(&all_64_pixels); // a screen frame

    screen.write_frame(&all_red_screen.frame_line()); // show the frame on the LED matrix
}
```


# Features

## `default` features

By default, the `basic`, and `linux-framebuffer` features are included.

## `basic` features

A set of features that don't require the hardware. This is mostly code that you will want to use if you are writing a simulator/emulator/etc. It includes, the `fonts`, `offset`, `rotate`, `clip`, `scroll`, and `serde-support` features.

### `fonts`

In `default`. A collection of legacy 8x8 fonts, renderable on the LED matrix.

### `offset`

In `default`. Support for offsetting the `PixelFrame` left/right/top/bottom. Requires `clip`.

### `rotate`

In `default`. Support for rotating `PixelFrame`s by 90-degree steps.

### `clip`

In `default`. Support for combining, and clipping two `PixelFrame`s onto a single frame.

### `scroll`

In `default`. Support for joining a collection of `PixelFrame`s into a single `Scroll`. Requires `clip`.

### `serde-support`

In `default`. Enables support for serialization/deserialization with `serde`.

## `linux-framebuffer`

In `default`. Use the Linux framebuffer to write to the LED matrix.

# Extra features

## `big-endian`

Uses big-endian format, suitable for non-AMD64/x86-64 processors. This is used when encoding/decoding 16-bit RGB565 to/from 24-bit RGB.
See [this](https://en.wikipedia.org/wiki/Endianness#Current_architectures) for more information.

Feature Wish List
=================
* [X] `linux-framebuffer` - In `default`. Use the Linux framebuffer to write to the LED matrix.
* [X] `fonts` - In `default`. A collection of legacy 8x8 fonts, renderable on the LED matrix.
* [X] `offset` - In `default`. Support for offsetting the `PixelFrame` left/right/up/down.
* [X] `rotate` - In `default`. Support for rotating `PixelFrame`s by 90-degree steps.
* [X] `clip` - In `default`. Support for combining, and clipping two `PixelFrame`s onto a single frame.
* [X] `scroll` - In `default`. Support for joining a collection of `PixelFrame`s into a single `Scroll`. Requires `clip`.
* [X] `serde-support` - In `default`. Enables support for serialization/deserialization with `serde`.
* [X] `big-endian` - Uses big-endian format, suitable for non-AMD64/x86-64 processors.

# Contribute

Please do contribute! Issues and pull requests are welcome.

Thank you for your help improving software one changelog at a time!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [v0.2.6] - 2021-01-25
### Changed
- Update font8x8 to v0.2.7.

## [v0.2.5] - 2020-06-17
### Changed
- Updated minimal rust version to v1.32.0
- Ran `cargo fix --edition`, with minor changes to `src/fram_clip.rs`,
  `src/frame_offset.rs`, and `src/frame_rotate.rs`.

### Added
- `edition = "2018"` under `[package]` in Cargo.toml.
- `1.32.0` toolchain to .travis.yml

## [v0.2.4] - 2020-06-15
### Changed
- Update framebuffer to v0.2.0.
- Run `cargo fmt` throughout.
- Removes `FontString::to_string` and implements `fmt::Display` as suggested
  by the `clippy::inherent_to_string_shadow_display` lint.

### Removed
- `rustfmt.toml`
- `clippy::into_iter_on_ref` lint.
- `clippy::len_without_is_empty` lint.
- `clippy::trivially_copy_pass_by_ref` lint.
- `clippy::inherent_to_string_shadow_display` lint. This removes `FontString::to_string`
  and implements `fmt::Display`.

## [v0.2.3] - 2019-06-07
### Changed
- Version for font8x8 to v0.2.5. Addresses https://github.com/rust-lang/rust/issues/61478

## [v0.2.2] - 2018-07-23
### Added
- New `error::ScreenError` type to handle crate errors.

### Fixed
- Avoid unnecessary allocations when initializing `FONT_COLLECTION`.

### Changed
- Use 'font8x8 = 0.2.3' with new `UnicodeFonts::iter` method.

## [v0.2.1] - 2018-07-20
### Fixed
- Correct version in Usage section of README.md

## [v0.2.0] - 2018-07-20
### Added
- `lazy_static = "1.0"` to Cargo.toml.
- Lazily-initiated static `fonts::FONT_HASHMAP` exported in top-level.
- Lazily-initiated static `fonts::FONT_COLLECTION` exported in top-level.

### Changed
- Update to `font8x8 = 0.2` in Cargo.toml.
- Replace `FontUtf16` with `FontUnicode` type.
- Replace `Utf16Fonts` with `UnicodeFonts` trait.
- Replace `u16` with `char`.
- Replace `FontString::encode_utf16` with `FontString::chars` method (breaking change).
- Update examples to use new Fonts API.
- `examples/letter.rs` uses `FONT_COLLECTION` instead of manually setting up a `FontCollection`.
- `examples/clip.rs` uses `FONT_COLLECTION` instead of manually setting up a `FontCollection`.

### Removed
- `fonts::print_collection` function.

## [v0.1.11] - 2018-06-17
### Fixed
- URLs to `examples/scroll-*.rs` on README.

## [v0.1.10] - 2018-06-17
### Added
- `BackgroundColor` trait defines setter/getter of background color.
- `StrokeColor` trait defines setter/getter of stroke color.
- Unit tests and documentation for:
  * `FontString::new`
  * `FontString::encode_utf16`
  * `FontString::to_string`
- `scroll` trait to join a collection of `PixelFrame`s into a single `Scroll`.
- `Scroll` type to wrap an inner `Vec<PixelFrame>`.
  * `Scroll::new(&[PixelFrame])` method initializes a scroll.
  * `Scroll::clips` method returns a `Vec<Clip>` made from the inner `Vec<PixelFrame>`.
  * `Scroll::frames` method returns a slice of the inner `Vec<PixelFrame>`.
  * `Scroll::reverse` method reverses the order of the inner `Vec<PixelFrame>`.
  * `Scroll::len` method the length of the inner `Vec<PixelFrame>`.
- Tests and implementation for new `FontFrame` type, which is a font with stroke and background colors:
  * `FontFrame::new`
  * `FontFrame::pixel_frame`
- `PixelFrame` implements `From<FontFrame>`.
- `FontFrame` implements `BackgroundColor` trait.
- `FontFrame` implements `StrokeColor` trait.
- New `basic` features are a subset of the current the `default` feature list.
- Add Travis CI integration with `.travis.yml`.
- Badge for Travis CI in README.md
- Add "Requirements" section to README.md.
- Specify Rust stable v1.26 as the minimum version requirement on README.md
- Tests for `Scroll` instances that create `FrameSequence` with a given `FrameDirection`.
  * `Scroll::right_to_left` for `FrameDirection::RightToLeft`.
  * `Scroll::left_to_right` for `FrameDirection::LeftToRight`.
  * `Scroll::top_to_bottom` for `FrameDirection::TopToBottom`.
  * `Scroll::bottom_to_top` for `FrameDirection::BottomToTop`.
- Add "Changelog" section to README.md, linking to this document.
- Add `FrameSequence` type.
- Add `examples/scroll-top-bottom.rs`.
- Add `examples/scroll-bottom-top.rs`.
- Add `examples/scroll-left-right.rs`.
- Add `examples/scroll-right-left.rs`.
- Add descriptions to examples on README.md.
- Add scroll examples to README.md

### Changed
- `FontString` wraps `Vec<font8x8::FontUtf16>` instead of `Vec<u16>`.
- Updated `examples/letter.rs` to use new `FontString::encode_utf16` method.
- Refactored `basic` features in the `default` feature list.
- `sensehat_screen::{Offset, clip_pixel_frames_offset_left, clip_pixel_frames_offset_top}` now have the `#[cfg(any(feature = "offset", feature = "clip"))]` attribute.

### Removed
- `rotate_bytes_left` and `rotate_bytes_right` implementation functions in `src/frame_clip.rs` are no longer needed.
- Duplicate badge for Cargo Docs from `README.md`.

## [v0.1.9] - 2018-05-01
### Added
- "Contribute" section in `CHANGELOG.md` to invite people to contribute to this library.
- `rustfmt.toml`, sets the expected `indent_style` to be `Visual`
- `impl<'a> From<&'a [PixelColor; 64]> for PixelFrame`
- `impl From<[PixelColor; 64]> for PixelFrame`
- `impl Into<[PixelColor; 64]> for PixelFrame`
- `impl Index<usize> for PixelFrame`, to get pixel colors using `let color = pixel_frame[idx]`-style syntax.
- `impl IndexMut<usize> for PixelFrame`, to set pixel colors using `pixel_frame[idx] = color`-style syntax.

### Changed
- Ran `cargo fmt` with `index_style` set to `Visual`

## [0.1.8] - 2018-04-30
### Fixed
- `font_to_frame` and `font_to_pixel_frame` methods were using incorrect indexes.

### Changed
- Update `README.md` section for examples, making it friendlier to read.

## [0.1.7] - 2018-04-29
### Added

src/lib.rs
//! A Rust library for the Raspberry Pi Sense HAT LED Screen
//! ========================================================
//!
//! The [Raspberry Pi Sense HAT](https://www.raspberrypi.org/products/sense-hat/) has an 8×8 RGB LED matrix that provides its own driver for the Linux framebuffer.
//!
//! This library provides a thread-safe, strong-typed, high-level API for the LED matrix, treating
//! it as you would any other screen on a Linux box.
//!
//! For examples, please check the
//! [examples/](https://github.com/saibatizoku/sensehat-screen-rs/tree/master/examples) folder in the source code.
//!
//! Basics
//! ------
//! * [`Screen`](./screen/struct.Screen.html) offers a high-level API for the LED Matrix.
//!
//!   Internally, it stores a `PixelFrame` meant to be rendered on the LED Matrix.
//!
//!   With the `linux-framebuffer` feature, enabled by default, `Screen` will have two methods:
//!
//!     1. `Screen::open` which opens the framebuffer
//!        file-descriptor given as the only argument.
//!
//!     1. `Screen::write_frame` which takes a
//!        `&FrameLine` and writes the raw bytes onto the framebuffer, effectively displaying the
//!        `PixelFrame` on the LED Matrix.
//!
//! * [`PixelFrame`](./frame/struct.PixelFrame.html) is a collection of 64 `PixelColor`, representing the 8-row by 8-column LED
//! Matrix.
//! * [`PixelColor`](./color/struct.PixelColor.html) is a 24-bit representation of an RGB color, encoded in three bytes.
//!
//! Low-level constructs
//! --------------------
//! * [`Rgb565`](./color/struct.Rgb565.html) is a 16-bit representation of an RGB color, encoded in two bytes. This is the
//! format. `Rgb565` converts into/from `PixelColor`.
//! supported by the LED Matrix's framebuffer device.
//! * [`FrameLine`](./frame/struct.FrameLine.html) is the raw-byte rendering of the `PixelFrame`,
//! properly encoded and ready to be written into the framebuffer device.
//!
//! Frame operations
//! ----------------
//! * [`Rotate`](./frame/rotate/enum.Rotate.html)
//!
//!   Requires `feature = "rotate"`, which is enabled by default.
//!
//!   Rotate the PixelFrame by `Rotate::None`, `Rotate::Ccw90`, `Rotate::Ccw180`, or
//!   `Rotate::Ccw270`, that correspond to a counter-clockwise angle of `0°`, `90°`, `180°`, and `270°`,
//!   respectively.
//!
//! * [`Offset`](./frame/offset/enum.Offset.html)
//!
//!   Requires `feature = "offset"`, which is enabled by default.
//!
//!   Offset the PixelFrame by `Offset::left(n)`, `Offset::right(n)`,
//!   `Offset::bottom(n)`, or `Offset::top(n)`, where `n` is an integer between in the `0..=8` range.
//!
//!   `Offset` with a value of `n = 0`, return a clone of the `PixelFrame`.
//!
//!   `Offset` with a value of `n = 8`, return a `PixelFrame` offset out of view, represented with black pixels (LEDs are off).
//!
//! * [`Clip`](./frame/clip/struct.Clip.html)
//!
//!   Requires `feature = "clip"`, which is enabled by default.
//!
//!   Creates a clip of two `PixelFrame`s, by defining an
//!   `Offset`. See the [clip documentation](./frame/clip/struct.Clip.html) for more details.
#[cfg(feature = "fonts")]
extern crate font8x8;
#[macro_use]
extern crate lazy_static;
#[cfg(feature = "linux-framebuffer")]
pub extern crate framebuffer;
#[cfg(feature = "serde-support")]
extern crate serde;
#[cfg(feature = "serde-support")]
#[macro_use]
extern crate serde_derive;

// RGB color with RGB565 support
pub mod color;
// Screen frames
pub mod frame;
// Screen errors
pub mod error;
// 8x8 fonts
#[cfg(feature = "fonts")]
pub mod fonts;
#[cfg(feature = "linux-framebuffer")]
#[path = "framebuffer.rs"]
pub mod screen;
// Scrolls for collections of PixelFrames
#[cfg(feature = "scroll")]
pub mod scroll;

// Re-exports
pub use self::color::{BackgroundColor, PixelColor, StrokeColor};

#[cfg(feature = "fonts")]
pub use self::fonts::{
    font_to_frame, font_to_pixel_frame, FontCollection, FontString, FONT_COLLECTION, FONT_HASHMAP,
};

#[cfg(feature = "clip")]
pub use self::frame::clip::Clip;

#[cfg(any(feature = "offset", feature = "clip"))]
pub use self::frame::Offset;

#[cfg(feature = "rotate")]
pub use self::frame::rotate::Rotate;

pub use self::frame::{FrameLine, PixelFrame};

#[cfg(feature = "linux-framebuffer")]
pub use self::screen::Screen;

#[cfg(feature = "scroll")]
pub use self::scroll::Scroll;
